kong/db/errors.lua

```diff
@@ -1033,13 +1032,14 @@ do
       for i, err_t_i in drain(section_errors) do
         local entity = entities[i]
 
-        if type(entity) == "table" then
+        if type(entity) == "table" and type(err_t_i) == "table" then
```

Comment:
I have an alternative suggestion. What if we coerce this specific type of error to an `@entity` error instead?

```suggestion
        -- promote error strings to `@entity` type errors
        if type(err_t_i) == "string" then
          err_t_i = { ["@entity"] = err_t_i }
        end

        if type(entity) == "table" and type(err_t_i) == "table" then
```

The goal of `/config?flatten_errors=1` is to flatten as much as possible into the `flattened_errors` array in the final output. The only things that should remain in the `fields` attribute are top-level errors such as an invalid/missing `_format_version`.

Here is the current response body:

```json
{
  "code": 14,
  "fields": {
    "consumers": [
      null,
      "uniqueness violation: 'consumers' entity with username set to 'test-consumer-1' already declared"
    ]
  },
  "flattened_errors": {},
  "message": "declarative config is invalid: {consumers={[2]=\"uniqueness violation: 'consumers' entity with username set to 'test-consumer-1' already declared\"}}",
  "name": "invalid declarative configuration"
}
```

If the suggestion above is applied, we get this instead:

```json
{
  "code": 14,
  "fields": {},
  "flattened_errors": [
    {
      "entity": {
        "id": "a73dc9a7-93df-584d-97c0-7f41a1bbce3d",
        "username": "test-consumer-1"
      },
      "entity_id": "a73dc9a7-93df-584d-97c0-7f41a1bbce3d",
      "entity_type": "consumer",
      "errors": [
        {
          "message": "uniqueness violation: 'consumers' entity with username set to 'test-consumer-1' already declared",
          "type": "entity"
        }
      ]
    }
  ],
  "message": "declarative config is invalid: {}",
  "name": "invalid declarative configuration"
}
```

It might be hard to see the usefulness of this extra verbosity and inclusion of the original entity that caused the error, but imagine if there was more data in the input:

```lua
    local input = {
      _format_version = "3.0",
      consumers = {
        {
          username = "test-consumer-1",
          tags = { "consumer-1-tag-1", "foo" },
        },
        {
          username = "test-consumer-1",
          tags = { "consumer-1-tag-2", "bar" },
        },
      },
    }
```

In this scenario, having the entire entity available adjacent to the error makes it clear what the offending entity is:

```json
{
  "code": 14,
  "fields": {},
  "flattened_errors": [
    {
      "entity": {
        "id": "a73dc9a7-93df-584d-97c0-7f41a1bbce3d",
        "tags": [
          "consumer-1-tag-2",
          "bar"
        ],
        "username": "test-consumer-1"
      },
      "entity_id": "a73dc9a7-93df-584d-97c0-7f41a1bbce3d",
      "entity_tags": [
        "consumer-1-tag-2",
        "bar"
      ],
      "entity_type": "consumer",
      "errors": [
        {
          "message": "uniqueness violation: 'consumers' entity with username set to 'test-consumer-1' already declared",
          "type": "entity"
        }
      ]
    }
  ],
  "message": "declarative config is invalid: {}",
  "name": "invalid declarative configuration"
}
```

Reply:
@programmer04 can you chime in here? I think that this suggestion makes sense, but you've originally reported this issue so I might be missing some color.

Reply:
The responses that the suggestion generates would be useful for KIC which uses the `flattened_errors` to associate affected entities with Kubernetes objects based on their tags. It allows us to find the affected entity easily as we get both the entity type and its id/name/tags and we do not have to depend on the order.

Reply:
> It allows us to find the affected entity easily

in fact the requirement is even stronger I think - if `flatten_errors` is enabled, Gateway **must** associate error messages with the entity if the error has been caused by a recognizable entity.

Reply:
We really only care about the tags in KIC--we can't do much of use with the Kong entity (users don't normally see those, they interact with the Kubernetes resource the tags identify).

I'm not clear on whether the suggested change would break anything in the existing schema. AFAICT that output looks like what we get with resources not affected by FTI-5584, which do list the error adjacent to the entity already: https://gist.github.com/rainest/34a47169c281090618df5d91f743a917

Is the suggestion just fixing some discrepancy between the fix and the normal flattened error output, or is it proposing a change to the flattened error format in general?

Reply:
The [FTI-5584](https://konghq.atlassian.net/browse/FTI-5584) has been reported an error in which ` flatten_errors=1` option can not flatten_error as expected. However, the FTI description wants the error message to be completely returned.

Yeah, ` flatten_errors=1`  option will parse the error message as instructed to another table,  But in the current case, `flatten_error` parses the error message has a type problem that causes flatten_error not to parse the error message expected, I have fixed it let parser does not process this type, return to the original error. That's the FTI author expected I think.

But @flrgh suggested me fix the parsing bug, let flattern_error successfully process this scenario, I think it's the best choice for this problem, but maybe it will break the FTI author something, this may confuse the KIC team, but after you confirm it, I will continue to make this change。

Reply:
@flrgh what you proposed should work as expected, but one question why it doesn't contain `entity_name` field?

Reply:
> why it doesn't contain `entity_name` field?

`entity_id`, `entity_tags`, and `entity_name` are copied from the input (`id`, `tags`, and `name` fields), so if they are not defined in the input, they will not be present in the final error structure